### PR TITLE
[scripts][forge] write fgi in python and edit job in place

### PR DESCRIPTION
**Pull request by rustielin**
## Motivation

Re-write `fgi` in python.

The motivation for python re-write:
* Shell script was useful for MVP, but an actual language would make it easier to maintain going forward
* vs rust, we really want a scripting language that's easy to run from a fresh environment. Python with its standard library is good for this.
* Easier access to kube API and fetching templates via JSON to parse
* Extensibility too. In the future if we see increased use of forge on k8s, we can implement cluster locking with s3 again. 

Aside from the re-write in a different language, there are a few notable differences:

* `fgi` edits the forge k8s job template rather than creating a singular pod whose lifecycle is managed completely by `fgi`. The job has a backoff limit of 0 for simplicity, meaning that it won't restart upon failure (this can probably be controlled programatically via `fgi` in the future, but would require some change in the way we determine a cluster's availability)
* Removed the ability to build and push images locally for now for simplicity. I don't want to add a new script, or to edit the existing one and share between `cti` and `fgi`. Currently `fgi` can build images from PR, or use existing image by tag.
* changed file paths, but otherwise maintain backwards compatibility of the commands/args. Command is invoked as `scripts/fgi/run` instead of `scripts/fgi`
* a fancy new ascii art!

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Run `fgi` with various options:

### Help

```
$ ./scripts/fgi/run --help
usage: run [-h] [--timeout-secs TIMEOUT_SECS] [--workspace WORKSPACE]
           [--env ENV] (--tag TAG | --pr PR | --local-swarm)

Entrypoint for the Forge unified testing framework

optional arguments:
  -h, --help            show this help message and exit
  --timeout-secs TIMEOUT_SECS
                        Timeout for the Forge pod in seconds
  --workspace WORKSPACE, -W WORKSPACE
                        Workspace or kubernetes cluster to run Forge on,
                        rather than picking one at random
  --env ENV, -E ENV     Extra environment variables to pass to Forge
  --tag TAG, -T TAG     Image tag to use in kubernetes Forge tests
  --pr PR, -p PR        PR to build images from for kubernetes Forge tests
  --local-swarm, -L     Run Forge tests locally instead of on a kubernetes
                        cluster
```

### Run with existing tag on kubernetes

```
$ time ./scripts/fgi/run -T dev_rustielin_pull_8674

    __________  ____  ____________
   / ____/ __ \/ __ \/ ____/ ____/
  / /_  / / / / /_/ / / __/ __/
 / __/ / /_/ / _, _/ /_/ / /___
/_/    \____/_/ |_|\____/_____/


Running Forge on backend: kubernetes testnet

Attempting to reach Forge Kubernetes testnets...
Updated context arn:aws:eks:us-west-2:853397791086:cluster/libra-rustietest in /Users/rustielin/.kube/config
Grabbing a testnet...
Running experiments on cluster: rustietest

Specfile: /var/folders/h2/5s1_h_yj343_wn3cg1_fx0yw0000gn/T/tmpmidk83mp.json
Creating job: forge-rustielin-1625095319
job.batch/forge-rustielin-1625095319 created
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
Waiting for forge-rustielin-1625095319 to be scheduled. Current phase: Pending
forge-rustielin-1625095319 reached phase: Running

**********
Auto refresh Dashboard: http://mon.rustietest.aws.hlw3truzy4ls.com/d/overview/overview?from=1625095348484&to=now&refresh=10s&orgId=1
**********

Log output: /var/folders/h2/5s1_h_yj343_wn3cg1_fx0yw0000gn/T/tmpv35m88y9
==========begin-pod-logs==========
deployment.apps/coredns scaled
Downloading and installing helm-s3 v0.10.0 ...
Checksum is valid.
Installed plugin: s3
"testnet-internal" has been added to your repositories
NAME                            CHART VERSION   APP VERSION     DESCRIPTION                                       
testnet-internal/diem-fullnode  1.3.0           1.3.0                                                             
testnet-internal/diem-validator 1.3.0           1.3.0           Diem blockchain validator                         
testnet-internal/testnet        0.1                             Additional components for a multi-validator tes...

running 4 tests
Starting to serve on 127.0.0.1:8001
Local kube pod healthcheck passed
test fund_account ... ok
test transfer_coins ... ok
test get_metadata ... ok
Will use 10 workers per endpoint with total 40 endpoint clients
Will create 15 accounts_per_client with total 600 accounts
Creating and minting faucet account
DD account current balances are 9223372020504717807, requested 600000000 coins
Completed minting seed accounts
Minting additional 600 accounts
Mint is done
starting emitting txns for 10 secs
Test Statistics: emit_transaction : 540 TPS, 1180 ms latency, 1450 ms p99 latency,no expired txns
test emit_transaction ... ok

test result: ok. 4 passed; 0 failed; 0 filtered out

==========end-pod-logs==========

**********
Dashboard snapshot: http://mon.rustietest.aws.hlw3truzy4ls.com/d/overview/overview?from=1625095348484&to=1625095386085&orgId=1
**********

Job forge-rustielin-1625095319 succeeded!
./scripts/fgi/run -T dev_rustielin_pull_8674  13.84s user 7.22s system 29% cpu 1:12.56 total
```

### Run with local swarm

```
$ ./scripts/fgi/run --local-swarm


    __________  ____  ____________
   / ____/ __ \/ __ \/ ____/ ____/
  / /_  / / / / /_/ / / __/ __/
 / __/ / /_/ / _, _/ /_/ / /___
/_/    \____/_/ |_|\____/_____/


Running Forge on local swarm backend
    Finished dev [unoptimized + debuginfo] target(s) in 0.74s
     Running `target/debug/forge --local-swarm`

running 5 tests
Wait for startup attempt: 0 of 10
rpc error: error sending request for url (http://localhost:58614/metrics): error trying to connect: tcp connect error: Connection refused (os error 61)
Wait for startup attempt: 1 of 10
Wait for connectivity attempt: 0
test fund_account ... ok
test transfer_coins ... ok
test get_metadata ... ok
Restarting node 652F27413456938EAA059F65231647CC
test restart_validator ... ok
Will use 10 workers per endpoint with total 10 endpoint clients
Will create 15 accounts_per_client with total 150 accounts
Creating and minting faucet account
DD account current balances are 9223372036854773807, requested 150000000 coins
Completed minting seed accounts
Minting additional 150 accounts
Mint is done
starting emitting txns for 10 secs
Test Statistics: emit_transaction : 15 TPS, 14884 ms latency, 14900 ms p99 latency,no expired txns
test emit_transaction ... ok

test result: ok. 5 passed; 0 failed; 0 filtered out
```

### CODE REVIEW

File: scripts/fgi/run [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            sys.exit(1)
+        subprocess.call(
+            ["./docker/build-aws.sh", "--build-forge", "--version", f"pull/{args.pr}"]
+        )
+        TAG = f"dev_{USER}_pull_{args.pr}"
+        print(
+            f"**TIP Use ./scripts/fgi -T {TAG} <...> to restart this run with the same tag without rebuilding it"
+        )
+
+# ================ Test setup ================
```
Comment (zihaoccc) [inline scripts/fgi/run:97]: I have to say: smoking hot nice drawing! Now we have our branding and big logo same as some luxury L and G words. hahah

**Review by bmwill — COMMENTED** [review]
Looks good to me (although it might be good to get someone who knows python better than me to review), and I agree with the simplifications you made, we can always revisit later if we need to.

File: scripts/fgi/kube.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        "image"
+    ] = f"{image_repo}:{tag}"
+    return job_name, template
+
+
+# ================ Kube queries ================
+
+
+def get_cluster_context(cluster_name):
+    k8s_context_pattern = "arn:aws:eks:us-west-2:853397791086:cluster/libra-CLUSTERNAME"
```
Comment (zihaoccc) [inline scripts/fgi/kube.py:72]: I think we did hide account number here in fgi?

Comment (rustielin) [reply scripts/fgi/kube.py]: updated to get it via the caller identity

**Review by zihaoccc — APPROVED** [review]

**Pull request closed without merging**